### PR TITLE
Discrepancies w/React Rendering

### LINKED ISSUE #83: Null components should be rendered
Consider the empty component:

    class Empty extends React.Component {
      render() {
        return null;
      }
    }

Then rendering the component should result in placeholder in the form of HTML comment like so

    > ReactDOMServer.renderToString(<Empty />)
    ''

Instead Rapscallion just outputs an empty string. Client-side React will then complain about invalid checksum.

This should also work with `undefined` and `false`.

### PR DESCRIPTION
**Pull request by ethersage**
This resolves a few differences between how react renders:

- `<!-- react-empty: {id} -->` for components that return null
- `<!-- react-text: {id} -->text<!-- /react-text>` is now rendered when there are multiple siblings to a text node
- False attribute values are not rendered: `disabled="false"`
- True attribute values are rendered without a value (like `disabled`), unless they are explicitly whitelisted attributes (`aria-haspopup` and `aria-expanded`)

Checksums appear to not be consistent with React either, seemingly because of a difference between `adler-32` and `react-dom/lib/adler32` but I'm not going to attempt to fix that here.

If this is acceptable I will write tests, update code comments, and clean lint.

Resolves #83

### CODE REVIEW

**Comment by ethersage** [comment]
@divmain Please let me know what you think.

**Comment by divmain** [comment]
This looks fantastic!  It's failing CI at the moment - feel free to add inline ESlint exceptions for everything but `valid-jsdoc`.  As soon as thats done, I'll merge this in.  Thanks!

File: src/render/attrs/index.js [hunk 1]
```diff
@@ -26,15 +31,33 @@ function renderAttrs (attrs) {
       !attrsNotToRender[attrKey]
     ) {
       let attrVal = attrs[attrKey];
+      const explicitBooleanValue = attrsWithExplicitBoolValue.includes(attrKey);
```
Comment (bmathews) [inline src/render/attrs/index.js]: Could `attrsWithExplicitBoolValue` be an object like `attrsNotToRender` to avoid the array search?

Comment (ethersage) [inline src/render/attrs/index.js]: Yes, good call.

**Comment by divmain** [comment]
It looks like this PR still needs a bit of work to pass CI.  Let me know if you want me to jump in at some point, and thanks for the work so far!

File: spec/special-cases.js [hunk 2]
```diff
@@ -9,7 +9,7 @@ describe("special cases", () => {
     const Parent = () => <div><NullComponent /></div>;
 
     return render(<Parent />).includeDataReactAttrs(false).toPromise().then(html => {
-      expect(html).to.equal("<div></div>");
+      expect(html).to.equal("<div><!-- react-empty --></div>");
```
Comment (mijamo) [inline spec/special-cases.js]: Shouldn't there be two tests here?

One that would be with `includeDataReactAttrs(false)` and that should return `"<div></div>"` and one that would be with `includeDataReactAttrs(true)` and that should return `"<div data-reactroot="" data-reactid="1"><!-- react-text: 1 --></div>"`

File: src/render/attrs/index.js [hunk 3]
```diff
@@ -10,6 +10,42 @@ const attrsNotToRender = {
   dangerouslySetInnerHTML: true
 };
 
+const attrsWithExplicitBoolValue = {
+  "aria-expanded": true,
+  "aria-haspopup": true
+};
```
Comment (mijamo) [inline src/render/attrs/index.js]: Just a thought here: there are many MANY properties and many cases. Maintaining that is a huge amount of work and it is very prone to errors. this is a potential major source of issues when comparing the react-dom version and rapscallion.

Wouldn't it be more reasonable to rely on the already crafted react internal accessible in the shared folder?

I was thinking we could maybe just have a custom function that would work a bit like https://github.com/facebook/react/blob/b1768b5a48d1f82e4ef4150e0036c5f846d3758a/src/renderers/dom/shared/DOMPropertyOperations.js#L138 that treats all the special cases handled by React currently and ensures that there is no difference when it comes to mutating arguments.

File: spec/special-cases.js (hunk 2 above)
Comment (ethersage) [inline spec/special-cases.js]: I'm not done with tests.

File: src/render/attrs/index.js (hunk 3 above)
Comment (ethersage) [inline src/render/attrs/index.js]: If there's something in React we can use for this logic, that sounds great. I don't quite understand the function you linked to. Could you show me how I might use it to replace my `attrsWithExplicitBoolValue` please?

I don't want to maintain a a list, but I don't currently understand what the logic is that would make the list unnecessary.

@mijamo

Comment (mijamo) [inline src/render/attrs/index.js]: I will try to make a quick POC this evening (being in Europe, this mean in the coming 6 hours) to explain it, as it is pretty hard to just describe it without code.

Comment (mijamo) [inline src/render/attrs/index.js]: Here I am, I took some time to make a working POC: 

https://gist.github.com/mijamo/9ece5c764071f343885401074e078182

The `style`property still needs to be handled separately, as seen in Fiber for instance (https://github.com/facebook/react/blob/b1768b5a48d1f82e4ef4150e0036c5f846d3758a/src/renderers/dom/fiber/ReactDOMFiberComponent.js#L335 ) but that simple POC should be able to handle all the other combinations of attr/values supported by React, as it corresponds to the internals of React as of today :)

Do not hesitate to ask if you need help integrating that in rapscallion, I can take a few hours from work to do that, we are very interested in rapscallion!

**Comment by ethersage** [comment]
@divmain Thanks, I'll let you know. I have time at work allocated to this so I should be able to get my tests done.

**Comment by ethersage** [comment]
Please don't merge this yet

File: src/render/attrs/index.js (hunk 3 above)
Comment (ethersage) [inline src/render/attrs/index.js]: @mijamo This is perfect, I will integrate today or tomorrow. That should be the last change on this PR I think @divmain

Comment (divmain) [inline src/render/attrs/index.js]: @mijamo, is `DOMProperty` considered a public API?

Comment (mijamo) [inline src/render/attrs/index.js]: Strictly speaking, the only public APIs of reactDomServer are `renderToString`  and `renderToStaticMarkup`. 

Relying on `DOMProperty` indeed has the downside of relying on some react internal that is subject to change. However, any change in the internal of React would most likely also affect the result of `renderToString` so Rapscallion needs to be tied with specific versions of React anyways to have a consistent behavior.

The good side is that it is in the `shared` folder and used by several different renderers, so it should not be ditched soon I believe.

Some other React libraries rely on it like https://github.com/chromakode/react-html-email for similar use case.

**Comment by ethersage** [comment]
@divmain 
@mijamo

**Comment by ethersage** [comment]
@divmain Actually hold on. Cheerio is causing some false positives on tests.

**Comment by ethersage** [comment]
@mijamo `DOMProperties.inject` is blowing up on tests. I tried checking for the existence of certain properties first but wasn't able to do it. I have no idea what's going on. Thoughts?

> Invariant Violation: injectDOMPropertyConfig(...): You're trying to inject DOM property 'aria-current' which has already been injected. You may be accidentally injecting the same DOM property config twice, or you may be injecting two configs that have conflicting property names.

**Comment by mijamo** [comment]
The issue is that `react-dom/server` does the injection on the same DOMProperty object when we import it in `spec/_utils`. So `react-dom` tries to do the injection and fails because we already did it ourselves before that.

A possible solution would be to import `react-dom/server` at the very beginning of the tests to ensure that they do their stuff with the injection, and then in Rpascallion code check for the properties od DOMProperty before we inject ourselves:

```
if (Object.keys(DOMProperty).length === 0) {
  DOMProperty.injection.injectDOMPropertyConfig(ARIADOMPropertyConfig);
  DOMProperty.injection.injectDOMPropertyConfig(HTMLDOMPropertyConfig);
  DOMProperty.injection.injectDOMPropertyConfig(SVGDOMPropertyConfig);
}
```

A cleaner solution would be to have our own DOMProperty separated from the react one to ensure there is no possible problem of interaction between the 2 of them. It could be simplified from the React DOMProperty because we don't need all the things done there (https://github.com/facebook/react/blob/15-stable/src/renderers/dom/shared/DOMProperty.js). A potential issue here is licensing as React is BSD and Rapscallion is MIT, but I guess that if we build our own DOMProperty and not just copy the react one there should not be any license issue?

I can try to build something in the coming hours and will share it afterwards in another gist.

**Comment by mijamo** [comment]
After some time trying different alternatives, it seems like importing `react-dom/server` at the beginning of the tests is really the easiest way, creating a homemade DOMProperty seems unnecessary and prone to errors.

**Comment by ethersage** [comment]
@divmain @mijamo If CI passes, this is ready 🎈

File: spec/special-cases.js [hunk 4]
```diff
@@ -12,6 +13,31 @@ describe("special cases", () => {
...
+    const TextWithSiblingsComponent = () => <div id="root"><div>child 1</div>child 2</div>;
+
+    const expected =
+      "<div data-reactid=\"2\">child 1</div><!-- react-text: 3 -->child 2<!-- /react-text -->";
+
+    return render(<TextWithSiblingsComponent />)
+      .toPromise()
+      .then(cheerio.load)
+      .then($ => $("#root").html())
+      .then(html => expect(html).to.equal(expected));
```
Comment (bmathews) [inline spec/special-cases.js]: What's cheerio doing for you here?

Comment (bmathews) [inline spec/special-cases.js]: Can you just include the wrapper div in the `expected` rather than bring in this dependency for this specific case?

Comment (ethersage) [inline spec/special-cases.js]: I could include the wrapper. The reason I didn't was because it has the checksum and seemed like a weird thing to embed into an unassociated unit test but it's not a big deal. I'll remove the dependency.

**Comment by ethersage** [comment]
@bmathews change made. All ready.

**Comment by ethersage** [comment]
@divmain Ready.

**Comment by fleg** [comment]
add failing test case for expression that returns null with siblings

**Comment by ethersage** [comment]
@fleg Merged your PR https://github.com/ethersage/rapscallion/pull/1

Empty nodes working and tests added.

**Comment by ethersage** [comment]
@jvivs Check this out.

@divmain

**Comment by fleg** [comment]
found one more bug with style

```jsx
  <div
    style={{
      margin: null
    }}
  />
```

```
      + expected - actual

      -<div style="" data-reactroot="" data-reactid="1" data-react-checksum="179180076"></div>
      +<div data-reactroot="" data-reactid="1" data-react-checksum="1998851930"></div>

```

**Comment by fleg** [comment]
more bugs I just found

multiple attribute in input
`<input multiple={false} />`

react: `<input/>`
rapscallion: `<input multiple="false"/>`

difference in attributes order
```
 (client) ></svg></div><input type="file" accept="image/jpeg,image/png" style="display:none;"
 (server) ></svg></div><input accept="image/jpeg,image/png" type="file" style="display:none;"
```

**Comment by divmain** [comment]
@fleg, could you file that as a separate issue?  There are a number of small discrepancies between React's rendering and Rapscallion's, so it's just going to be a matter of documenting the differences and knocking them out, one at a time.

**Comment by divmain** [comment]
There was a small conflict that I resolved locally.  This now in master.

**Comment by getkey** [comment]
Could you publish this to NPM? :-)

**Comment by divmain** [comment]
@getkey, there is a batch of PRs that will be going in over the next day or two.  Feel free to pester me if you don't see a release by Thursday :)

**Comment by getkey** [comment]
Awesome, thanks!

**Comment by indeyets** [comment]
@divmain ping! waiting for release too :)

**Comment by getkey** [comment]
@divmain ping! I'd love to see these PRs merged and published, this lib is really nice and would improve the experience of our users but currently this is what prevents me to implement it. :-)

**Comment by divmain** [comment]
Thanks @getkey and @indeyets :) Was waiting on one other PR to land, and it has!  I just pushed out `v2.1.8`.

Thanks for your patience!  As a side note, I've been working on (another) side project that I hope will help me manage the maintenance of my OSS projects.  I hope that'll improve responsiveness and streamline the process of marshaling PRs through review. 🤞

**Comment by indeyets** [comment]
@divmain thanks! please consider adding [release notes](https://github.com/FormidableLabs/rapscallion/releases)

this version is a huge advancement

**Comment by divmain** [comment]
My not-quite-ready OSS management project will help to automate project releases, including more helpful release data.  I'm working on it :)

**Comment by indeyets** [comment]
@divmain sorry for polluting this PR. release seems to be broken.

`rapscallion/lib/renderer.js:10` requires `"adler-32"` which is not a dependency anymore

**Comment by divmain** [comment]
Damn. Thanks for the heads up.

**Comment by divmain** [comment]
It was a build issue.  I've fixed with a `prepublish` `package.json` entry and pushed `2.1.9` that should install correctly.

**Comment by indeyets** [comment]
@divmain it is related to 'react-dom' imports in components
